### ISSUE TITLE
Recurrence problem

### ISSUE THREAD

**Issue by ASmallWhit** [zx33/APAUNet#4]
I have read through your paper. I now expect to reproduce the code with the pancreatic tumor data of task 7 in MSD, preprocess the data according to the requirements in readme, and put it into the specified folder, and start training. But the training results differ by 9 percentage points from the accuracy written in the paper. Did you train according to the parameters already written in the existing code at that time? Thank you very much.

**Comment by zx33** [comment]
This code was indeed refactored when uploaded to github, but we have tested it running consistently.
Do you directly using the 'evaluation' function in 'trainer.py' to obtain the dice score? That function is just a trick for quick evaluation on validation set which only evaluates **ONE** (96, 96, 96) bounding cropped patch from the whole scan for each sample. You should use the 'slide_windows_inference' in MONAI for evaluation.

Anyway, I'll test this code in the next few days again to confirm whether there is a problem.

**Comment by ASmallWhit** [comment]
Thank you very much for your reply. According to your method, I replaced the "evaluation" in trainer.py with the "slide_windows_conference" in MONAI for evaluation, but the score of dice is still quite different. It may also be that there is a problem with my replacement method. Could you please upload the code of "evaluation" in trainer.py with the "slide_windows_conference" in MONAI for evaluation to github? Thank you very much




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月24日(星期五) 晚上8:53
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





 
This code was indeed refactored when uploaded to github, but we have tested it running consistently.
 Do you directly using the 'evaluation' function in 'trainer.py' to obtain the dice score? That function is just a trick for quick evaluation on validation set which only evaluates ONE (96, 96, 96) bounding cropped patch from the whole scan for each sample. You should use the 'slide_windows_inference' in MONAI for evaluation.
 
Anyway, I'll test this code in the next few days again to confirm whether there is a problem.

**Comment by zx33** [comment]
Hi,

I've upload the 'test.py' for evaluation. I'm also double-checking this code, if there is any problem, I will fix it to GitHub.

**Comment by ASmallWhit** [comment]
Ok, I'll train and test the accuracy again. Thank you very much for your answer and help. I wish you a happy life

**Comment by ASmallWhit** [comment]
Ok, I'll train and test the accuracy again. Thank you very much for your answer and help. I wish you a happy life


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月25日(星期六) 下午4:11
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





 
Hi,
 
I've upload the 'test.py' for evaluation. I'm also double-checking this code, if there is any problem, I will fix it to GitHub.

**Comment by ASmallWhit** [comment]
Hello, about the code you resubmitted to github, I retrained it again. Now the dice score of pancreatic organs has exceeded 1, and the average dice score of pancreatic organs and pancreatic tumors has reached more than 0.8, which does not correspond to the results in your paper. Do I feel that the newly submitted method of calculating dice score is unreasonable? Could you have a look again when you have time? I also hope that you can retrain the code in github and test how to achieve the accuracy in the paper. Thank you very much.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月25日(星期六) 下午4:11
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





 
Hi,
 
I've upload the 'test.py' for evaluation. I'm also double-checking this code, if there is any problem, I will fix it to GitHub.

**Comment by zx33** [comment]
Hi,

I find that I miss the inter in the denominator in the calculation of dice.

Of course I'll the check these codes again. Thank you for your feedback about this repository.

**Comment by sweetsweet-hub** [comment]
Hey buddy! Can you please be extra careful? I'm really counting on you for my code! Thank you so much~

**Comment by ASmallWhit** [comment]
> Hey buddy! Can you please be extra careful? I'm really counting on you for my code! Thank you so much~

Hello, I am retraining and testing this code. I will give feedback in time if there are any follow-up questions

**Comment by ASmallWhit** [comment]
Hello, I'm very sorry. According to your newly uploaded code, after I used MSD pancreatic tumor segmentation data to retrain 800 epochs, the dice score of pancreatic tumor with set accuracy verified by test.py test is only 46%, which is still quite different from the 55% segmentation accuracy of pancreatic tumor in your paper. Can you use this code to retrain and test whether the segmentation of pancreatic tumor can reach the accuracy in your paper? We look forward to your reply and improvement. Thank you very much for your contribution.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月26日(星期天) 上午10:45
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





 
Hi,
 
I find that I miss the inter in the denominator in the calculation of dice.
 
Of course I'll the check these codes again. Thank you for your feedback about this repository.

**Comment by zx33** [comment]
Yes, I'm re-running this code and checking what went wrong right now.
But it is during the DDL period right now, and the GPU resources (this experiment needs gpu with large memory, e.g., A100) need to be queued for a long time, thus the process of debugging may not be that fast.

> Hello, I'm very sorry. According to your newly uploaded code, after I used MSD pancreatic tumor segmentation data to retrain 800 epochs, the dice score of pancreatic tumor with set accuracy verified by test.py test is only 46%, which is still quite different from the 55% segmentation accuracy of pancreatic tumor in your paper. Can you use this code to retrain and test whether the segmentation of pancreatic tumor can reach the accuracy in your paper? We look forward to your reply and improvement. Thank you very much for your contribution.

**Comment by ASmallWhit** [comment]
> 是的，我正在重新运行此代码并检查现在出了什么问题。但是现在正值 DDL 期间，GPU 资源（这个实验需要大内存的 GPU，比如 A100）需要排队很长时间，因此调试的过程可能不会那么快。
> 
> > 您好，非常抱歉。根据您新上传的代码，在我使用MSD胰腺肿瘤分割数据重新训练800个epoch后，经 test.py 测试验证的胰腺肿瘤的骰子得分只有46%，这与您论文中胰腺肿瘤的55%分割准确率仍然有很大差异。你能用这段代码来重新训练和测试胰腺肿瘤的分割是否能达到你论文中的准确性吗？我们期待您的回复和改进。非常感谢您的贡献。

Ok, thank you very much for your reply and look forward to your further correction.

**Comment by ASmallWhit** [comment]
Hello, have you verified the accuracy of the code? Because I have seen so many codes, I think your code framework is the best, so can you please verify it as soon as possible? Thank you very much.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年2月27日(星期一) 上午10:21
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





 
Yes, I'm re-running this code and checking what went wrong right now.
 But it is during the DDL period right now, and the GPU resources (this experiment needs gpu with large memory, e.g., A100) need to be queued for a long time, thus the process of debugging may not be that fast.
  
Hello, I'm very sorry. According to your newly uploaded code, after I used MSD pancreatic tumor segmentation data to retrain 800 epochs, the dice score of pancreatic tumor with set accuracy verified by test.py test is only 46%, which is still quite different from the 55% segmentation accuracy of pancreatic tumor in your paper. Can you use this code to retrain and test whether the segmentation of pancreatic tumor can reach the accuracy in your paper? We look forward to your reply and improvement. Thank you very much for your contribution.

**Comment by zx33** [comment]
> Hello, have you verified the accuracy of the code? Because I have seen so many codes, I think your code framework is the best, so can you please verify it as soon as possible? Thank you very much.

Hello, we'll update as soon as this weekend or next week if there's any progress.
PS: the DDL is at midnight of March 9th (AoE), we'll have enough computing resources to start working on this after that.
Thanks for your reply and patience.

**Comment by ASmallWhit** [comment]
> Yes, I'm re-running this code and checking what went wrong right now. But it is during the DDL period right now, and the GPU resources (this experiment needs gpu with large memory, e.g., A100) need to be queued for a long time, thus the process of debugging may not be that fast.
> 
> > Hello, I'm very sorry. According to your newly uploaded code, after I used MSD pancreatic tumor segmentation data to retrain 800 epochs, the dice score of pancreatic tumor with set accuracy verified by test.py test is only 46%, which is still quite different from the 55% segmentation accuracy of pancreatic tumor in your paper. Can you use this code to retrain and test whether the segmentation of pancreatic tumor can reach the accuracy in your paper? We look forward to your reply and improvement. Thank you very much for your contribution.

Hello, have you verified the accuracy of the code? Because I have seen so many codes, I think your code framework is the best, so can you please verify it as soon as possible? Thank you very much.

> Hello. The code framework they provided is really good. After my testing, there were no issues, so it was easy for me to replace it with my own model and I found that the accuracy surpassed the highest reported level in their paper. I did not test their own model, so I am not sure how it performs.
Hello brother, is it convenient to send me the code you have trained that is higher than the accuracy of the paper? I don't know why I changed several models and still couldn't get up to the standard. Maybe there was a problem with the way I changed them. Could you send me your code and let me see how you changed it? Thank you very much.

If you can, you can package the code and send it to my mailbox: fywlbzy@163.com

**Comment by ASmallWhit** [comment]
Ok, thank you very much for your reply and look forward to your further update.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年3月5日(星期天) 中午1:07
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





  
Hello, have you verified the accuracy of the code? Because I have seen so many codes, I think your code framework is the best, so can you please verify it as soon as possible? Thank you very much.
  
Hello, we'll update as soon as this weekend or next week if there's any progress.
 PS: the DDL is at midnight of March 9th (AoE), we'll have enough computing resources to start working on this after that.
 Thanks for your reply and patience.

**Comment by ASmallWhit** [comment]
Hello, I use the weight I have trained now to run conference.py to predict the result graph, and find that 80G of video memory is not enough. How much memory does conference.py need to run?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年3月5日(星期天) 中午1:07
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





  
Hello, have you verified the accuracy of the code? Because I have seen so many codes, I think your code framework is the best, so can you please verify it as soon as possible? Thank you very much.
  
Hello, we'll update as soon as this weekend or next week if there's any progress.
 PS: the DDL is at midnight of March 9th (AoE), we'll have enough computing resources to start working on this after that.
 Thanks for your reply and patience.

**Comment by zx33** [comment]
> Hello, I use the weight I have trained now to run conference.py to predict the result graph, and find that 80G of video memory is not enough. How much memory does conference.py need to run?

Oh, the `torch.no_grad()` is missed in `inference.py`
The `test.py` inference is the correct way
```
# original:
result = monai.inferers.sliding_window_inference(sample, (96, 96, 96), 4, model, overlap=0.5, cval=-3)
# replace:
with torch.no_grad():
  result = monai.inferers.sliding_window_inference(sample, (96, 96, 96), 4, model, overlap=0.5, cval=-3)
```

**Comment by ASmallWhit** [comment]
Hello, after your suggestion, the conference.py file can be run, but the size of the predicted result image is far from the original image, and the segmented pancreas and tumor area can not be seen. I feel that there are many problems in the code training process, the calculation method of dice score, the test accuracy and the use of test sets for reasoning. I hope you can verify and correct it as soon as you have time, Thank you very much.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年3月6日(星期一) 上午10:44
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





  
Hello, I use the weight I have trained now to run conference.py to predict the result graph, and find that 80G of video memory is not enough. How much memory does conference.py need to run?
  
Oh, the torch.no_grad() is missed in inference.py
 The test.py inference is the correct way
 # original: result = monai.inferers.sliding_window_inference(sample, (96, 96, 96), 4, model, overlap=0.5, cval=-3) # replace: with torch.no_grad():   result = monai.inferers.sliding_window_inference(sample, (96, 96, 96), 4, model, overlap=0.5, cval=-3)

**Comment by ASmallWhit** [comment]
> > 您好，您是否验证了代码的准确性？因为我看过这么多代码，我觉得你们的代码框架是最好的，所以你能不能尽快验证一下？谢谢。
> 
> 您好，如果有任何进展，我们将在本周末或下周更新。 PS：DDL 在 9 月 <> 日午夜（AoE），之后我们将有足够的计算资源开始处理。感谢您的回复和耐心等待。

Hello, I'm sorry to bother you again. Have you verified this code?

**Comment by zx33** [comment]
Yes, we are currently working on this.

**Comment by ASmallWhit** [comment]
> Yes, we are currently working on this.

Hello, have you verified it?

**Comment by ASmallWhit** [comment]
Hello, have you finished the verification?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年3月12日(星期天) 上午8:57
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





 
Yes, we are currently working on this.

**Comment by zx33** [comment]
> Hello, have you finished the verification?

Yes, my collaborator is currently investigating this issue, and if there is any progress, he will update it here: https://github.com/Oliver-jiang/APAUNet.

**Comment by ASmallWhit** [comment]
> > Hello, have you finished the verification?
> 
> Yes, my collaborator is currently investigating this issue, and if there is any progress, he will update it here: https://github.com/Oliver-jiang/APAU

**Comment by ASmallWhit** [comment]
Hello, have you verified the accuracy issue here? Can I add your WeChat and ask you some code questions?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zx33/APAUNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年3月18日(星期六) 上午8:48
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zx33/APAUNet] Recurrence problem (Issue #4)





  
Hello, have you finished the verification?
  
Yes, my collaborator is currently investigating this issue, and if there is any progress, he will update it here: https://github.com/Oliver-jiang/APAUNet.

**Comment by ztz123321** [comment]
I found the same problem while running in your code，I have tried training your network and other different networks under your code, and found that none of them can predict the segmentation results.  Looking  forward to your verification

**Comment by zx33** [comment]
> I found the same problem while running in your code，I have tried training your network and other different networks under your code, and found that none of them can predict the segmentation results. Looking forward to your verification

Sure, this code base was refactored by me, and some implementations may contain some issues. 
My collaborator is checking about this and he will update the code on his github once completed.

**Comment by zx33** [comment]
> Can I add your WeChat and ask you some code questions?

You can email your questions to me and I will reply as I see the mail.
